lms/djangoapps/bulk_email/models.py

```diff
@@ -52,29 +50,113 @@ class Meta(object):
         abstract = True
 
 
+# Bulk email targets - the send to options that users can select from when they send email.
+SEND_TO_MYSELF = 'myself'
+SEND_TO_STAFF = 'staff'
+SEND_TO_LEARNERS = 'learners'
+SEND_TO_COHORT = 'cohort'
+SEND_TO_ALL = 'all'
+EMAIL_TARGETS = [SEND_TO_MYSELF, SEND_TO_STAFF, SEND_TO_LEARNERS, SEND_TO_COHORT, SEND_TO_ALL]
+EMAIL_TARGET_DESCRIPTIONS = ['Myself', 'Staff and instructors', 'All students', 'Specific cohort', 'All']
+EMAIL_TARGET_CHOICES = zip(EMAIL_TARGETS, EMAIL_TARGET_DESCRIPTIONS)
+
+
+class Target(models.Model):
+    """
+    A way to refer to a particular group (within a course) as a "Send to:" target.
+    """
+    target_type = models.CharField(max_length=64, choices=EMAIL_TARGET_CHOICES)
+
+    class Meta(object):
+        app_label = "bulk_email"
+
+    def __unicode__(self):
+        return "CourseEmail Target for: {}".format(self.target_type)
+
+    def get_users(self, course_id, user_id=None):
+        """
+        Gets the users for a given target.
+
+        Result is returned in the form of a queryset, and may contain duplicates.
+        """
+        staff_qset = CourseStaffRole(course_id).users_with_role()
+        instructor_qset = CourseInstructorRole(course_id).users_with_role()
+        staff_instructor_qset = (staff_qset | instructor_qset)
+        enrollment_qset = User.objects.filter(
+            is_active=True,
+            courseenrollment__course_id=course_id,
+            courseenrollment__is_active=True
+        )
+        if self.target_type == SEND_TO_MYSELF:
+            if user_id is None:
+                raise ValueError("Must define self user to send email to self.")
+            user = User.objects.filter(id=user_id)
+            return use_read_replica_if_available(user)
+        elif self.target_type == SEND_TO_STAFF:
+            return use_read_replica_if_available(staff_instructor_qset)
+        elif self.target_type == SEND_TO_LEARNERS:
+            if staff_instructor_qset.count() <= 0:
+                return use_read_replica_if_available(enrollment_qset)
+            return use_read_replica_if_available(enrollment_qset.exclude(id__in=staff_instructor_qset))
+        elif self.target_type == SEND_TO_COHORT:
+            return User.objects.none()  # TODO: cohorts aren't hooked up, put that logic here
+        elif self.target_type == SEND_TO_ALL:
```

Comment:
Do we wish to keep SEND_TO_ALL? If so, some of the test changes below can be simplified.